### ISSUE TITLE
Segmentation fault running layers.conv2d

### ISSUE THREAD

**Issue by wschoong** [tensorflow/tensorflow#22885]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Centos 7.5
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: NA
- **TensorFlow installed from (source or binary)**: Anaconda
- **TensorFlow version (use command below)**: 1.10.0
- **Python version**: 3.6.5
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**: python test.py

### Describe the problem
I am trying to implement a CNN using tensorflow, but I started running into segmentation fault with increasing input size to layers.conv2d. In the provided simple test script, I can successfully run it if n_H=3000. But I got a segmentation fault when I set n_H=4000. In addition, if I run it without layars.conv2d by setting with_conv=False, the script runs successfully. I don't know if this is my lack of understanding with layers.conv2d or there is something more to this.

### Source code / logs
Here is the output when I get a segmentation fault:
```
$ python test.py
(100, 4000, 100, 1) (100, 8)
2018-10-10 11:57:23.825704: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2
2018-10-10 11:57:23.827653: I tensorflow/core/common_runtime/process_util.cc:69] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
Fatal Python error: Segmentation fault

Thread 0x00007fea9af1a740 (most recent call first):
  File "/home/seng/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1350 in _call_tf_sessionrun
  File "/home/seng/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1263 in _run_fn
  File "/home/seng/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1278 in _do_call
  File "/home/seng/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1272 in _do_run
  File "/home/seng/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1100 in _run
  File "/home/seng/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 877 in run
  File "test.py", line 43 in <module>
Segmentation fault (core dumped)
```

Here is the simple test script test.py:
```
import faulthandler; faulthandler.enable()
import numpy as np
import tensorflow as tf

seed = 42
tf.reset_default_graph()
tf.set_random_seed(seed)
np.random.seed(seed)

num_examples = 100
n_H = 4000
with_conv = True

# generate training data
X_train = np.random.randn(num_examples*n_H*100).reshape(num_examples, n_H, 100, 1)
Y_train = np.random.randn(num_examples*8).reshape(num_examples, 8)    
print(X_train.shape, Y_train.shape)

# create placeholders
X = tf.placeholder(tf.float32, shape=(None, n_H, 100, 1))
Y = tf.placeholder(tf.float32, shape=(None, 8))

# build graph
if (with_conv):
    conv1 = tf.layers.conv2d(X, filters=64, kernel_size=[5, 5],strides = 1, padding='valid',activation = tf.nn.relu)    
    pool1 = tf.layers.max_pooling2d(conv1, pool_size=[2, 2], strides=2, padding='valid')
else:
    pool1 = tf.layers.max_pooling2d(X, pool_size=[2, 2], strides=2, padding='valid')    
pool1_flat = tf.layers.flatten(pool1)
dense2 = tf.layers.dense(pool1_flat, units=256, activation=tf.nn.relu)
H = tf.layers.dense(dense2, units=8, activation=tf.nn.relu)

# compute cost
cost = tf.reduce_mean(tf.square(Y - H))

# initialize variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    a = sess.run(cost, feed_dict = {X: X_train, Y:Y_train})
    print(a)

```

**Comment by wschoong** [comment]
Initially I thought I was saturating my RAM. I have 45 GB and my host has 12 CPUs. During the execution, my test script used up about 20 GB before segmentation fault. What system memory configuration could affect the execution? How would I fix my system memory configuration? Have you try increasing n_H to larger values (6000, 8000, etc)? I can also get segmentation fault by increasing the num_examples.

**Comment by wschoong** [comment]
Just an update. I just ran my script on an OS X host with 32 GB RAM and tensorflow version. It ran successfully!
 One thing I notice is there was no message stating like I had on my Centos 7 host:
`2018-10-10 11:57:23.827653: I tensorflow/core/common_runtime/process_util.cc:69] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.`
Why does it segmentation fault on my host running Centos 7?

**Comment by ymodak** [comment]
I was able to run your script successfully on increasing the n_H value till 9000. Going beyond that landed me in segmentation fault similar to yours. The tensorflow/core/framework/allocator.cc script does the calculation of system memory usage. This script must allocate less memory for computation for your Centos 7 host system config than the threshold memory.

**Comment by wschoong** [comment]
Does this mean allocator.cc incorrectly calculated the system memory usage for Centos 7?

**Comment by ymodak** [comment]
No. It calculated correctly however the memory allocation was not good enough to execute test.py in Centos & system configuration.

**Comment by ymodak** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)

**Comment by wschoong** [comment]
I don't really know how to resolve my problem. I understand this may not be a tensorflow issue. But how do I increase the memory allocation? If insufficient memory is allocated, shouldn't an exception be thrown? Any help would be appreciated.

**Comment by ymodak** [comment]
@asimshankar  Can you please take a look at this issue and suggest how this should be addressed?

**Comment by asimshankar** [comment]
@wschoong : Thanks for reporting. Yes, it is running out of memory, and yes it should fail gracefully instead of segfaulting.

I haven't traced when, but I _think_ this has been fixed, though after 1.11.0, so the fix will only be visible in 1.12.0. I tried the same program out in 1.10.0, 1.11.0, and 1.12.0-rc0 and while I see a segmentation fault in the first two (as you guys did), I see a better error message:

```
ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[15353856,256] and type float on /job:localhost/replica:0/task:0/device:CPU:0 by allocator cpu
         [[node dense/kernel/Initializer/random_uniform/RandomUniform (defined at ./test.py:29)  = RandomUniform[T=DT_INT32, _class=["loc:@dense/kernel/Assign"], dtype=DT_FLOAT, seed=42, seed2=32, _device="/job:localhost/replica:0/task:0/device:CPU:0"](dense/kernel/Initializer/random_uniform/shape)]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.
```

in 1.12.0-rc0.

Closing this out since it has been fixed since, and we're unlikely to want to patch previous releases.
Feel free to reopen if you think I'm mistaken.

Thanks!
